File: LeetCode/Problems/C++/78. Subsets.cpp
//bit manipulation
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        if(!nums.size()) return {};
        vector<vector<int>> ret;
        int bit_count;
        bit_count = 1 << nums.size();
        for(int i=0;i<bit_count;i++){
            vector<int> cur;
            for(int j =0;j<nums.size();j++){
                if ((i>>j) & 1)
                    cur.push_back(nums[j]);
            }
            ret.push_back(cur);
        }
        return ret;
    }
};

//cascading
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        if(!nums.size()) return {};
        vector<vector<int>> ret;
        //subset(nums, 0, ret, {});
        ret.push_back({});
        for(int i=0;i< nums.size();i++){
            int result = ret.size();
            for(int j=0;j<result;j++){
                vector<int> cur =ret[j];
                cur.push_back(nums[i]);
                ret.push_back(cur);
            }
        }
        return ret;
    }
};

//backtracking
class Solution {
public:
    void subset(vector<int> &nums, int idx, vector<vector<int>>& ret, vector<int> curr, int target){
        if(curr.size() == target){
            ret.push_back(curr);
            return;
        }

        for(int i = idx; i< nums.size(); i++){
            curr.push_back(nums[i]);
            subset(nums, i+1, ret, curr, target);
            curr.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        if(!nums.size()) return {};
        vector<vector<int>> ret;
        for(int k=0;k<nums.size()+1;k++)
            subset(nums, 0, ret, {}, k);
   
        return ret;
    }
};

File: LeetCode/Problems/Python/78. Subsets.py
#bit manipulation
#Time: O(n * 2^n)
#Space: (n * 2^n)
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        output = []
        
        for i in range(2**n, 2**(n + 1)):
            # generate bitmask, from 0..00 to 1..11
            #print(bin(i))
            bitmask = bin(i)[3:]
            output.append([nums[j] for j in range(n) if bitmask[j] == '1'])
            
        return output
#cascading
#Time: O(n * 2^n)
#Space: (n * 2^n)
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        output = [[]]
        for i in range(n):
            size = len(output)
            for j in range(size):
                cur = output[j].copy()
                cur.append(nums[i])
                output.append(cur)
        return output

        
#backtracking
#Time: O(n * 2^n)
#Space: (n * 2^n)class Solution:
    def backtracking(self, nums, idx, curr, target, output):
        if len(curr) == target:
            output.append(curr)
            
        for i in range(idx, len(nums)):
            self.backtracking(nums, i+1, curr+[nums[i]], target, output)
            
    def subsets(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        
        n = len(nums)
        output = []
        for i in range(n+1):
            self.backtracking(nums, 0, [], i, output)
        return output